### PR TITLE
Kma 161770415 update rate engine shipmentlineitem

### PR DESCRIPTION
**Pull request by kimallen**
## Description

This modifies the rate engine's `ComputeShipment` so that it returns not just the fee with the cost but the rate (with discount rate applied) and then saves it to the shipment_line_items table with the other values.  

It also modifies the "ComputeShipmentLineItemCharge" so that it returns the discounted rate as well.  This rate gets  saved to shipment_line_items when the pre-approvals are saved.

Stores the applied_rate as millicents instead of cents.

## Reviewer Notes

Are there any missing tests or error handling?  

Note that the linehaul rate should have a value of zero (since it represents a total of multiple fees, not a value based on one rate)

## Setup

To test in the app, you will need to get a shipment to the "in transit" state, making sure the shipment has a net weight and an actual pickup date. You should be able to populate your DB with the db_populate_e2e script, then do a local login to the TSP interface with tspuser1@example.com. Go to the "in transit" shipments in the queue, then pick the shipment with locator ENTDEL. Click the Deliver button and submit a delivery date. You should now see applied rates for records in the ShipmentLineItem table.

Changes value of applied_rate in shipment_line_items table to be in millicents

## Code Review Verification Steps

* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/161770415) for this change

### CODE REVIEW

**Comment by kimallen** [comment]
I realized I need to add some code after Patrick's PR lands in order to make sure the applied_rate gets saved in the tables for the accessorials, so changing this to WIP.

**Comment by kimallen** [comment]
No longer WIP

File: cmd/demo/rateengine.go [hunk 1]
```diff
@@ -82,10 +82,10 @@ func main() {
...
 		fmt.Printf("%-30s%s\n", "Linehaul total (trans. cost):", cost.LinehaulChargeTotal.ToDollarString())
 		fmt.Println("")
-		fmt.Printf("%-30s%s\n", "Origin service fee:", cost.OriginServiceFee.ToDollarString())
-		fmt.Printf("%-30s%s\n", "Destination service fee:", cost.DestinationServiceFee.ToDollarString())
-		fmt.Printf("%-30s%s\n", "Full Pack fee:", cost.PackFee.ToDollarString())
-		fmt.Printf("%-30s%s\n", "Full Unpack fee:", cost.UnpackFee.ToDollarString())
+		fmt.Printf("%-30s%s\n", "Origin service fee:", cost.OriginService.Fee.ToDollarString())
+		fmt.Printf("%-30s%s\n", "Destination service fee:", cost.DestinationService.Fee.ToDollarString())
+		fmt.Printf("%-30s%s\n", "Full Pack fee:", cost.Pack.Fee.ToDollarString())
+		fmt.Printf("%-30s%s\n", "Full Unpack fee:", cost.Unpack.Fee.ToDollarString())
```
[the remark this answers is not in the document]
Comment (kimallen) [reply cmd/demo/rateengine.go]: Sure- adding

File: pkg/rateengine/accessorials.go [hunk 2]
```diff
@@ -104,12 +104,12 @@ var tariff400ngWeightBasedItems = map[string]bool{
 }
 
 // ComputeShipmentLineItemCharge calculates the total charge for a supplied shipment line item
-func (re *RateEngine) ComputeShipmentLineItemCharge(shipmentLineItem models.ShipmentLineItem) (unit.Cents, error) {
+func (re *RateEngine) ComputeShipmentLineItemCharge(shipmentLineItem models.ShipmentLineItem) (unit.Cents, unit.Cents, error) {
```
[the remark this answers is not in the document]
Comment (kimallen) [reply pkg/rateengine/accessorials.go]: Using the FeeAndRate struct does seem to be more readable to understand what those values represent.  For the accessorials, the value of the returned rate equals the discounted rate; whereas in the nonlinehaul struct it represents the non-discounted rate.  Would it be confusing to use the same struct for each?

File: pkg/rateengine/line_items.go [hunk 3]
```diff
@@ -10,6 +10,7 @@ import (
 // CreateBaseShipmentLineItems will create and return the models for the base shipment line items that every
 // shipment should contain
 func CreateBaseShipmentLineItems(db *pop.Connection, costByShipment CostByShipment) ([]models.ShipmentLineItem, error) {
+	// TODO: add AppliedRate to each nonLinehaul item from cost
```
[the remark this answers is not in the document]
Comment (kimallen) [reply pkg/rateengine/line_items.go]: Removing

File: pkg/rateengine/nonlinehaul_test.go [hunk 4]
```diff
@@ -33,14 +33,14 @@ func (suite *RateEngineSuite) Test_CheckServiceFee() {
...
+	feeAndRate, err := engine.serviceFeeCents(unit.CWT(50), "395", testdatagen.DateInsidePeakRateCycle)
 	if err != nil {
 		t.Fatalf("failed to calculate service fee: %s", err)
 	}
 
 	expected := unit.Cents(17500)
-	if fee != expected {
-		t.Errorf("wrong service fee: expected %d, got %d", expected, fee)
+	if feeAndRate.Fee != expected {
+		t.Errorf("wrong service fee: expected %d, got %d", expected, feeAndRate.Fee)
```
[the remark this answers is not in the document]
Comment (kimallen) [reply pkg/rateengine/nonlinehaul_test.go]: Yes, adding those tests

File: pkg/rateengine/scenario_test.go [hunk 5]
```diff
@@ -30,8 +30,8 @@ func (suite *RateEngineSuite) Test_Scenario1() {
 	suite.Assertions.Nil(err, "could not compute PPM")
 
 	suite.Equal(unit.Cents(163434), cost.LinehaulChargeTotal)
-	suite.Equal(unit.Cents(4765), cost.OriginServiceFee)
-	suite.Equal(unit.Cents(5689), cost.DestinationServiceFee)
+	suite.Equal(unit.Cents(4765), cost.OriginService.Fee)
+	suite.Equal(unit.Cents(5689), cost.DestinationService.Fee)
```
[the remark this answers is not in the document]
Comment (kimallen) [reply pkg/rateengine/scenario_test.go]: I don't think so for these tests.  They are testing the PPM, which doesn't require a returned rate value.  But let me know if you think it still should.

File: pkg/rateengine/accessorials.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikena-truss) [reply pkg/rateengine/accessorials.go]: +1 to using a struct here. Maybe just describe the return in the doc string so the two don't get mixed up.

File: pkg/rateengine/scenario_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (reggieriser) [reply pkg/rateengine/scenario_test.go]: OK, sounds good.  I missed that this was for PPMs.

File: pkg/rateengine/line_items.go [hunk 6]
```diff
@@ -90,6 +98,7 @@ func CreateBaseShipmentLineItems(db *pop.Connection, costByShipment CostByShipme
 		Quantity1:         bqNetWeight,
 		Status:            models.ShipmentLineItemStatusSUBMITTED,
 		AmountCents:       &packAndUnpackFee,
+		AppliedRate:       &packRate,
 		SubmittedDate:     time.Now(),
```
Comment (mikena-truss) [inline pkg/rateengine/line_items.go]: Out of scope of your changes, but it looks like this is the only line item that recomputes `time.Now()`. Should this match the rest of them above?

**Comment by kimallen** [comment]
applied_rate is now represented as millicents in order to represent the accuracy of 'Full Unpack' whose rate is providedin millicents

File: pkg/rateengine/line_items.go (hunk 6 above)
Comment (reggieriser) [reply pkg/rateengine/line_items.go]: Good catch -- this was something I probably missed in a refactor earlier.  @kimallen , do you mind changing `time.Now()` to `now` here?

File: swagger/api.yaml [hunk 7]
```diff
@@ -291,7 +291,10 @@ definitions:
         example: 10000
       applied_rate:
         type: integer
+        format: cents
         minimum: 0
+        x-nullable: true
+        description: unit is cents
```
Comment (reggieriser) [inline swagger/api.yaml]: Change `description` and `format` above to millicents?

File: pkg/unit/cents.go [hunk 8]
```diff
@@ -50,3 +53,25 @@ func (c Cents) ToDollarFloat() float64 {
...
+func (c Millicents) MultiplyFloat64(f float64) Millicents {
+	return Millicents(math.Round(float64(c.Int()) * f))
+}
+
+// ToDollarString returns a dollar string representation of this value
+func (c Millicents) ToDollarString() string {
+	d := float64(c) / 100000.0
+	s := fmt.Sprintf("$%.2f", d)
+	return s
+}
```
Comment (reggieriser) [inline pkg/unit/cents.go]: I was initially thinking we'd add millicents in its own millicents.go file like the other units.  But cents/millicents are related in a way the other units aren't.  Not sure one way is better than the other necessarily.

**Pull request merged**